library/lua/script-data.lua

```diff
@@ -0,0 +1,344 @@
+end
+
+-- Sets the autosave flag for an entry
+function set_autosave(id, do_autosave)
+  get_loaded_data(id).autosave = do_autosave
+end
+
+-- Set whether the current save should save world entries when quitting
+-- This value is reset back to false whenever a world is unloaded / loaded
+-- It's automatically set to true whenever an arena game is started
+function set_quit_without_saving(bool)
+  quit_without_saving = bool
+end
+
+-- Returns the loaded data entry (essentially the metadata of a loaded script data) for the script data of the given ID
+-- There's not likely a reason to use this externally.
+function get_loaded_data(id)
+  return loaded_data[id]
+end
+
+-- Check to see if save path info is recorded
+-- If it's not, attempt to update the record if possible
+function update_save_path()
+  if recorded_save_path == nil and dfhack.isWorldLoaded() then
+    recorded_save_path = dfhack.getSavePath()
+  end
+end
+---------------------------------------------------------------------
+-- MISC
+
+local function world_load_setup()
+  -- We store the path of the loaded save just in case there's some hiccups with the fact we try to save world data during the SC_WORLD_UNLOADED event (where dfhack.getSavePath() might possibly return nil)
+  recorded_save_path = dfhack.getSavePath()
+
+  -- Begin monitoring for autosaves / quicksaves
+  -- Since autosaves are only actually relevant in fortress mode, there's no reason to start checking for them outside of that mode
+  if dfhack.world.isFortressMode() then
+    start_loop()
+  end
+
+  -- Automatically disable saving on exit if in arena mode
+  -- This is because it's the only current mode(?) that allows for quitting without saving, and there's no way (without some hackery) to really detect that
+  -- DFhack will also report it as having a save directory that doesn't actually exist if the arena game hasn't been saved, just to add to the annoyances
+  if dfhack.world.isArena() then
+    set_quit_without_saving(true)
+  else
+    set_quit_without_saving(false)
+  end
+end
+
+dfhack.onStateChange["script-data"] = function(event)
+  if event == SC_WORLD_LOADED then
+    world_load_setup()
+
+  elseif event == SC_WORLD_UNLOADED then
+    -- Trigger the saves for all `world` entries, provided quitting without saving is disabled
+    if quit_without_saving == false then
+      for id, entry in pairs(loaded_data) do
+        if entry.type == "world" then
+          save(id)
+        end
+      end
+    end
+
+    -- Now, cleanup world-related data (while still keeping global data around)
+    for id, entry in pairs(loaded_data) do
+      if entry.type == "world" then
+        loaded_data[id] = nil
+      end
+    end
+
+    recorded_save_path = nil
+    set_quit_without_saving(false)
+
+    -- Stop monitoring autosaves
+    if loop_timer ~= nil then
+      dfhack.timeout_active(loop_timer, nil)
+    end
+  end
+end
+
+function start_loop()
+  loop_timer = dfhack.timeout(1, "frames", autosave_check_loop)
```

Comment:
I'm tempted to say that we should hold off on this until we have a solution for #1733. But I'll test it as-is for now, and I'll try to prioritize a proper solution there.